src/ffi/array.rs

```diff
@@ -241,12 +235,21 @@ unsafe fn create_buffer<T: NativeType>(
         return Ok(Buffer::new());
     }
 
-    let ptr = get_buffer_ptr(array, data_type, index)?;
-
     let offset = buffer_offset(array, data_type, index);
-    let bytes = Bytes::from_foreign(ptr, len, BytesAllocator::InternalArrowArray(owner));
+    let ptr: *mut T = get_buffer_ptr(array, data_type, index)?;
 
-    Ok(Buffer::from_bytes(bytes).sliced(offset, len - offset))
+    // We have to check alignment.
+    // This is the zero-copy path.
+    if ptr.align_offset(std::mem::align_of::<T>()) == 0 {
+        let bytes = Bytes::from_foreign(ptr, len, BytesAllocator::InternalArrowArray(owner));
```

Comment:
Bytes takes ownership of `InternalArrowArray(owner)` ensuring the memory backed by the pointer is not released.